akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonManager.scala

```diff
@@ -683,7 +731,80 @@ class ClusterSingletonManager(
       self ! DelayedMemberRemoved(m)
   }
 
-  def gotoOldest(): State = {
+  // Try and go to oldest, taking the lease if needed
+  def tryGoToOldest(): State = {
+    // check if lease
+    lease match {
+      case None ⇒
+        goToOldest()
+      case Some(_) ⇒
+        logInfo("Trying to acquire lease before starting singleton")
+        tryAcquireLease()
+    }
+  }
+
+  when(AcquiringLease) {
+    case Event(AcquireLeaseResult(result), _) ⇒
+      logInfo("Acquire lease result {}", result)
+      if (result) {
+        goToOldest()
+      } else {
+        setTimer(LeaseRetryTimer, LeaseRetry, leaseRetryInterval)
+        stay using AcquiringLeaseData(leaseRequestInProgress = false, None)
+      }
+    case Event(Terminated(ref), AcquiringLeaseData(_, Some(singleton))) if ref == singleton ⇒
+      logInfo("Singleton actor terminated. Trying to acquire lease again before re-creating.")
+      tryAcquireLease()
+    case Event(AcquireLeaseFailure(t), _) ⇒
+      log.error(t, "failed to get lease (will be retried)")
+      setTimer(LeaseRetryTimer, LeaseRetry, leaseRetryInterval)
+      stay using AcquiringLeaseData(leaseRequestInProgress = false, None)
+    case Event(LeaseRetry, _) ⇒
+      tryAcquireLease()
+    case Event(OldestChanged(oldestOption), _) ⇒
+      oldestChangedReceived = true
+      logInfo("AcquiringLease observed OldestChanged: [{} -> {}]", cluster.selfAddress, oldestOption.map(_.address))
+      oldestOption match {
+        case Some(a) if a == cluster.selfUniqueAddress ⇒
+          // already oldest
+          stay
+        case Some(a) if !selfExited && removed.contains(a) ⇒
+          handOverDone(None)
+        case Some(a) ⇒
+          // send TakeOver request in case the new oldest doesn't know previous oldest
+          peer(a.address) ! TakeOverFromMe
+          setTimer(TakeOverRetryTimer, TakeOverRetry(1), handOverRetryInterval, repeat = false)
+          goto(WasOldest) using WasOldestData(None, newOldestOption = Some(a))
+        case None ⇒
+          // new oldest will initiate the hand-over
+          setTimer(TakeOverRetryTimer, TakeOverRetry(1), handOverRetryInterval, repeat = false)
+          goto(WasOldest) using WasOldestData(None, newOldestOption = None)
+      }
+    case Event(HandOverToMe, _) ⇒
+      gotoHandingOver(None, Some(sender()))
```

Comment:
why not use the singleton Option from AcquiringLeaseData?

Reply:
Doh, originally there couldn't be a singleton actor in this state, missed this one when adding it